### PR TITLE
suit: in-place updateable components

### PR DESCRIPTION
**Pull request by SylwesterKonczyk**

### CODE REVIEW

**Review by tomchy — APPROVED** [review]

File: subsys/suit/platform/sdfw/src/suit_plat_ipuc.c [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+	k_mutex_lock(&ipuc_mutex, K_FOREVER);
+	for (size_t i = 0; i < CONFIG_SUIT_IPUC_SIZE; i++) {
+		suit_plat_ipuc_entry_t *p_entry = &components[i];
+
+		if (allocated_entry == NULL && p_entry->component_id.value == NULL) {
+			allocated_entry = p_entry;
+
+		} else if (suit_compare_zcbor_strings(component_id, &p_entry->component_id)) {
+			allocated_entry = p_entry;
```
Comment (ahasztag) [inline subsys/suit/platform/sdfw/src/suit_plat_ipuc.c:101]: Hard to say without seeing the usage of `write_offset`, but what if `p_entry->write_offset` is not 0 at this stage? Isn't this a result of a malformed manifest?

File: subsys/suit/platform/sdfw/src/suit_plat_ipuc.c [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+			suit_component_type_t component_type = SUIT_COMPONENT_TYPE_UNSUPPORTED;
+
+			intptr_t slot_address = 0;
+			size_t slot_size = 0;
+
+			if (suit_plat_decode_component_type(component_id, &component_type) !=
+				    SUIT_PLAT_SUCCESS ||
+			    component_type != SUIT_COMPONENT_TYPE_MEM) {
+				continue;
```
Comment (ahasztag) [inline subsys/suit/platform/sdfw/src/suit_plat_ipuc.c:149]: Looking at `suit_plat_ipuc_declare` (more specifically at `ipuc_component_id_from_handle`) a component of type different then `MEM` should never land in the `components` array. Shouldn't we return an error here?

File: subsys/suit/platform/sdfw/src/suit_plat_copy.c [hunk 3]
```diff
@@ -204,6 +208,10 @@ int suit_plat_copy(suit_component_t dst_handle, suit_component_t src_handle,
 	 * Stream the data.
 	 */
 
+#if CONFIG_SUIT_IPUC
+	suit_plat_ipuc_revoke(dst_handle);
```
[the remark this answers is not in the document]
Comment (SylwesterKonczyk) [reply subsys/suit/platform/sdfw/src/suit_plat_copy.c:212]: Functions like suit_plat_copy() are not 'aware' if given component is in-place updateable at given moment.

so meaning of suit_plat_ipuc_revoke() is: "If given component is declared as in-place updateable, then such declaration shall be removed now"

File: subsys/suit/platform/sdfw/src/suit_plat_ipuc.c (hunk 1 above)
Comment (SylwesterKonczyk) [reply subsys/suit/platform/sdfw/src/suit_plat_ipuc.c:101]: It should not be a problem here, as write_offset will be updateable as result of IPC execution, which should not occur while manifest processing is still in progress

File: subsys/suit/platform/sdfw/src/suit_plat_ipuc.c (hunk 2 above)
Comment (SylwesterKonczyk) [reply subsys/suit/platform/sdfw/src/suit_plat_ipuc.c:149]: No, IPUC will have two use cases:

1. using its space as 'sdfw mirror', i.e. when SDFW candidate is placed in external flash
2. updating inactive component from appliaction.

1st case indeed can operate on MEM component only, but 2nd case may be supported on other component types as well

**Review by ahasztag — APPROVED** [review]

File: subsys/suit/platform/sdfw/Kconfig [hunk 4]
```diff
@@ -22,6 +22,20 @@ config SUIT_DIGEST_CACHE_SIZE
 
 endif
 
+config SUIT_IPUC
+	bool "Enable in-place updateable components"
+	help
+		SUIT manifest may instruct the platform that component is inactive, by overriding
+		image-size parameter with value 0. Such component may be updated in place,
+		also memory associated to such component may be utilized for other purposes.
```
Comment (nordicjm) [inline subsys/suit/platform/sdfw/Kconfig:30]: help text indent is 1x tab followed by 2x spaces

**Review by nordicjm — CHANGES_REQUESTED** [review]
minor nit.

File: subsys/suit/platform/sdfw/Kconfig (hunk 4 above)
Comment (tomchy) [reply subsys/suit/platform/sdfw/Kconfig]: Fixed!

**Review by nordicjm — APPROVED** [review]

**Pull request merged**